#10330 uses non-3.12 CMake features

When I implemented https://github.com/cvc5/cvc5/pull/10330, I used:

```
SOURCE_FROM_CONTENT
```

Not realising this was new in CMake 3.25.

As such, `cvc5` fails to build with CMake 3.12, which is the documented minimum for the project.

This was reported by @BrunoDutertre here: https://github.com/cvc5/cvc5/commit/f77908cdb196f75fbc63f7c3d05ea35a1e4c7f44#commitcomment-139003709

Here's the error message:
```
CMake Error: The source directory "CaDiCaL_version.cpp" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
CMake Error at cmake/FindCaDiCaL.cmake:27 (try_run):
  Failed to configure test project build system.
Call Stack (most recent call first):
  CMakeLists.txt:445 (find_package)


CMake Error: TRY_COMPILE attempt to remove -rf directory that does not contain CMakeTmp:SOURCE_FROM_CONTENT
```

--- a/cmake/FindCaDiCaL.cmake
+++ b/cmake/FindCaDiCaL.cmake
@@ -23,10 +23,10 @@ find_library(CaDiCaL_LIBRARIES NAMES cadical)
 
 set(CaDiCaL_FOUND_SYSTEM FALSE)
 if(CaDiCaL_INCLUDE_DIR AND CaDiCaL_LIBRARIES)
-  # Try to compile our version file
-  try_run(RUN_RESULT_VAR
-    COMPILE_RESULT_VAR
-    SOURCE_FROM_CONTENT CaDiCaL_version.cpp
+
+  # Generate our version check file in CMAKE_BINARY_DIR
+  set(CaDiCaL_version_src "${CMAKE_BINARY_DIR}/CaDiCaL_version.cpp")
+  file(WRITE ${CaDiCaL_version_src}
     "
     #include <cadical.hpp>
     #include <iostream>
@@ -37,9 +37,28 @@ if(CaDiCaL_INCLUDE_DIR AND CaDiCaL_LIBRARIES)
       return 0;
     }
     "
+  )
+
+  # `try_run` doesn't have a way to specify a specific include dirs, so we need
+  # to set (and then reset) `CMAKE_CXX_FLAGS`; our version file is a .cpp file,
+  # so CXX_FLAGS are the right flags
+  set(OLD_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+
+  # cvc5 doesn't support MSVC, so we're fine to hard-code `-I` as the include
+  # flag
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I${CaDiCaL_INCLUDE_DIR}")
+
+  # Try to compile and run our version file
+  try_run(RUN_RESULT_VAR
+    COMPILE_RESULT_VAR
+    ${CMAKE_BINARY_DIR}
+    ${CaDiCaL_version_src}
     LINK_LIBRARIES ${CaDiCaL_LIBRARIES}
     RUN_OUTPUT_VARIABLE CaDiCaL_VERSION
   )
+
+  # Restore our CXX flags after checking with `try_run`
+  set(CMAKE_CXX_FLAGS ${OLD_CXX_FLAGS})
 
   # If this failed to compile or run, we have bigger issues
   if (NOT ${RUN_RESULT_VAR} EQUAL 0 OR NOT ${COMPILE_RESULT_VAR})

